classes/mail/Mailable.php

```diff
@@ -451,23 +455,43 @@ public static function getDataDescriptions(): array
             }
         }
 
-        foreach ($args as $arg) { /** @var ReflectionParameter $arg) */
-            $argClassName = $arg->getType()->getName();
-
+        foreach ($args as $arg) {
+            $argTypes = static::getTypeNames($arg->getType());
             foreach ($map as $dataClass => $variableClass) {
-                if (is_a($argClassName, $dataClass, true)) {
-                    $descriptions = array_merge(
-                        $descriptions,
-                        $variableClass::descriptions()
-                    );
-                    continue 2;
+                foreach ($argTypes as $argType) {
+                    if (is_a($argType, $dataClass, true)) {
+                        $descriptions = array_merge(
+                            $descriptions,
+                            $variableClass::descriptions()
+                        );
+                        // An intersection type could be passed in (e.g. Submission & PublishedSubmission), so allow the code check other maps for correctness
+                        continue 2;
+                    }
                 }
             }
         }
 
         return $descriptions;
     }
 
+    /**
+     * Retrieve the list of type names that might compose a given type
+     * @return string[]
+     */
+    protected static function getTypeNames(ReflectionType $type): array
+    {
+        if ($type instanceof ReflectionNamedType) {
+            return [$type->getName()];
+        }
+        if ($type instanceof ReflectionUnionType || $type instanceof ReflectionIntersectionType) {
```

Comment:
On the top of my head, Mailables can support intersection types but not union types, except cases, like in ORCID Plugin, where classes belong to different applications (but nevertheless, I'd check if it works as expected)

Constructor arguments are identified before Mailable instantiation to forecast what email template variables are going to be available when sending an email. E.g., say class1 is a associated with email template variable1 and class2 with variable2, then using the union
```php
public function __construct(class1|class2 $object)
```   
 would mean that both template variables, 1 and 2, are available, which isn't true.  

As a workaround we can ensure that both class names in the union don't exist in the application at the same time and exclude the one that isn't defined. Or forbid union types and deal with it on the plugin side. Or maybe there is a cleaner way...

Reply:
Ops, I forgot to answer 😅

I thought about this case, and it will work out of the box for our main case `Server|Journal|Preprint` (only one of the classes will exist, and the `is_a()` will skip the other variants properly).

Reply:
For the other cases (where the class might actually exist), like let's say an email that might receive a `Failure|Success`... I think we have the following options:
- A. Build a separate email class for each variant
  This is the simplest for us and users, so that would be my choice. It will become unmaintainable just if we had many of such "intersection" variables in a single email, but then we can revisit this topic.
- B. Add "Smarty" capabilities to the template, so the users can setup loops/conditions/etc (`We {if Failure)failed{else}succeeded{/if}`)
  This would require adding a "view code" / building a powerful and intuitive editor 😅

Reply:
So, if you don't have anything against, I think it's ok to merge it this way, and use the option A 🤔

Reply:
And for future updates, I think that maybe it's better to have a method/interface, like errrrrr, `public static function getVariables()`, then we can call it and the class will still have a chance to customize its variables.

Reply:
@jonasraoni, thanks for double checking that it works. Can you also add an exception to unsure that both classes in the union don't exist at the same time? It's easy to forget it, especially for someone who haven't worked with mailables before.

Reply:
When there's more than one active class in the union type, I've updated the code to retrieve only the first class and emit a warning.
Anyway, I think it's worth to open an issue to revisit how these variables are retrieved later.